indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionStrategy.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.task;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.druid.indexer.TaskStatus;
+import org.apache.druid.indexing.common.TaskToolbox;
+import org.apache.druid.java.util.common.NonnullPair;
+import org.apache.druid.segment.indexing.DataSchema;
+import org.joda.time.Interval;
+
+import java.util.List;
+
+/**
+ * Strategy to be used for executing a compaction task.
+ * Should be synchronized with {@link org.apache.druid.server.coordinator.UserCompactionStrategy}
+ */
+@JsonTypeInfo(
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.EXISTING_PROPERTY,
+    property = CompactionStrategy.TYPE_PROPERTY,
+    visible = true)
+@JsonSubTypes(value = {
+    @JsonSubTypes.Type(name = NativeCompactionStrategy.TYPE, value = NativeCompactionStrategy.class)
+})
+public interface CompactionStrategy
+{
+  String TYPE_PROPERTY = "TYPE";
+
+  TaskStatus runCompactionTasks(
+      CompactionTask compactionTask,
+      TaskToolbox taskToolbox,
+      List<NonnullPair<Interval, DataSchema>> dataSchemas
+  ) throws JsonProcessingException;
+
+  CurrentSubTaskHolder getCurrentSubTaskHolder();
```

Comment:
This interface should also have a supportsCompactionSpec() method which will tell the run task method can it run the compaction spec using this strategy or not. 
This will be helpful once we fallback to native in case msq is set to default as a cluster level engine. 
If we donot have it, MSQ engine would have to support 100% feature coverage of native compaction spec, else users jobs will fail.

Reply:
Added `supportsCompactionConfig` at both `ClientCompactionRunnerInfo` and `CompactionRunner`. Yes we were validating things such as `PartitionSpecs` type and `maxTotalRows` being null, apart from other MSQ context flag checks introduced now.